### ISSUE TITLE
Flag significant differences in AF/FAF across v2.1 exomes and v3.1 genomes (and update on v4)

### ISSUE THREAD

**Issue by gtiao** [broadinstitute/gnomad_qc#190]

**Comment by gtiao** [comment]
Council would like us to flag if FAFs are significantly different across exomes and genomes. They also noted Elizabeth Atkinson did work a few years ago identifying variants with significantly different AFs across v2.1 exomes and genomes, and we will invite her to present her work at an upcoming project meeting.

Council would like us to do this for:
- v2.1 exomes and v3.1 genomes
- v4 exomes and v3.1 (i.e., v4) genomes

Then feed this information back to exome capture designers.

PIs note that the hard part will be determining WHY this is happening, and we don't have much bandwidth for this at the moment. Also noted that ancestry represented could be different (not well matched) across exomes and genomes, so we should flag differences on pop-specific FAFs

**Comment by jkgoodrich** [comment]
Cluster:
```
hailctl dataproc start jg1 --properties=spark:spark.speculation=true --autoscaling-policy=max-100 --max-idle=100m --requester-pays-allow-all --pkgs="git+https://github.com/broadinstitute/gnomad_methods.git@master"
```

Command:
```
hailctl dataproc submit jg1 PycharmProjects/gnomad_qc/gnomad_qc/v3/annotations/compare_freq.py  --overwrite --pyfiles PycharmProjects/gnomad_qc/gnomad_qc
```

Job Link: https://console.cloud.google.com/dataproc/jobs/075676ecd8dd4bea9fed33fa39cd7b33/monitoring?region=us-central1&project=broad-mpg-gnomad

**Comment by jkgoodrich** [comment]
For reference, here are Elizabeth Atkinson's ATGU lab meeting slides that have the v2 exome/genome allele frequency discrepancy plots

[https://app.zenhub.com/files/153521178/1919c8b4-39e0-427a-a459-512b21252da5/download](https://app.zenhub.com/files/153521178/1919c8b4-39e0-427a-a459-512b21252da5/download)

**Comment by jkgoodrich** [comment]
As a first step, I remade the plots that Elizabeth made (not subset to bi-allelic variants). All plots are filtered to variants with AC > 5 within the given population for both exomes and genomes.

**v3 genomes - v2 exomes deltaAF:**
![bokeh_plot - 2021-11-30T155527.616.png](https://images.zenhubusercontent.com/5f4d3fb5900ef2def7626452/a812702e-a4c0-4a08-8877-d921e1518fa7)
![bokeh_plot - 2021-11-30T155531.205.png](https://images.zenhubusercontent.com/5f4d3fb5900ef2def7626452/5bd3ad61-e786-4cc1-a2d1-f8ded3a5133c)
![bokeh_plot - 2021-11-30T155534.381.png](https://images.zenhubusercontent.com/5f4d3fb5900ef2def7626452/dc8af5c7-85db-405f-b4f6-42c77cb104b2)
![bokeh_plot - 2021-11-30T155538.399.png](https://images.zenhubusercontent.com/5f4d3fb5900ef2def7626452/06766932-9aef-443c-b920-018b94e669c1)
![bokeh_plot - 2021-11-30T155541.280.png](https://images.zenhubusercontent.com/5f4d3fb5900ef2def7626452/7db1b046-efb9-49ae-81de-bb37dffcdb5b)
![bokeh_plot - 2021-11-30T155543.614.png](https://images.zenhubusercontent.com/5f4d3fb5900ef2def7626452/adcc1e37-aecf-42d9-8c4f-eb7f3534b2d4)
![bokeh_plot - 2021-11-30T155545.466.png](https://images.zenhubusercontent.com/5f4d3fb5900ef2def7626452/9789c324-49a9-4407-b694-b4e7fda65c3a)
![bokeh_plot - 2021-11-30T155547.486.png](https://images.zenhubusercontent.com/5f4d3fb5900ef2def7626452/56bf1c85-6723-4715-80c3-a9bef051afe2)


**chi squared p-value comparing v3 genomes to v2 exomes AF**
![bokeh_plot - 2021-11-30T155555.635.png](https://images.zenhubusercontent.com/5f4d3fb5900ef2def7626452/448a37a0-f2ce-4838-9f00-2f5d0062b27d)
![bokeh_plot - 2021-11-30T155558.896.png](https://images.zenhubusercontent.com/5f4d3fb5900ef2def7626452/079f5df7-cfb5-4311-b89e-6bbffa33c190)
![bokeh_plot - 2021-11-30T155603.100.png](https://images.zenhubusercontent.com/5f4d3fb5900ef2def7626452/879567ea-dacd-4c8e-bcab-78edd979fc34)
![bokeh_plot - 2021-11-30T155605.547.png](https://images.zenhubusercontent.com/5f4d3fb5900ef2def7626452/1e310097-a0ef-443c-9c8d-b04cd35c0ba0)
![bokeh_plot - 2021-11-30T155607.847.png](https://images.zenhubusercontent.com/5f4d3fb5900ef2def7626452/723ccdcc-f274-4a22-8523-23de2c7dd337)
![bokeh_plot - 2021-11-30T155610.230.png](https://images.zenhubusercontent.com/5f4d3fb5900ef2def7626452/58822912-47b2-46d9-bfc2-a090ad4a7ba8)
![bokeh_plot - 2021-11-30T155612.381.png](https://images.zenhubusercontent.com/5f4d3fb5900ef2def7626452/103987bc-8fdf-420e-9b0f-7898d906e910)
![bokeh_plot - 2021-11-30T155614.315.png](https://images.zenhubusercontent.com/5f4d3fb5900ef2def7626452/0f88dedc-a208-449b-9739-3667f37e2a57)


So if we want to add an indication of some sort warning users that there are notable AF differences between exomes and genomes, we need to decide exactly what cutoffs to use for this.

I few things to think about:
- Plots above are AC > 5 , how should we decide on the cutoff to use here, how do we justify that decision?
- Plots above are by population, do we indicate if any population significantly differs, or only indicate for each population where this is true?
- What p-value cutoff should we use and how should we justify that decision?
- Should we look at differences by variant type?

**Comment by jkgoodrich** [comment]
After discussions at the MPG/DSP meeting we plan to:

- [ ] Try an AN cutoff instead of the AC of 5
- [ ] Read over Elizabeth's manuscript on this https://docs.google.com/document/d/1Hmyt5dK5BBG6PbaIowGZSObgnbIMT7uB5lHN4465DkY/edit?pli=1#heading=h.tk7pmofqozeu 
- [ ] Apply suggestions from the manuscript to get numbers of variants that would be filtered
- [ ] Assess this stratified by variant type: bi-allelic/multi-allelic, SNV/indel
- [ ] Present results at gnomAD project meeting on Jan 4th

**Comment by jkgoodrich** [comment]
Results of the number of variants that would be excluded with the suggested p-value cutoff in Elizabeth's paper: <1x10-5 https://docs.google.com/spreadsheets/d/1HtP6t2_ta4EkcyU4PIh-c9c0fUz1mhVBSlcEnC_gUqw/edit#gid=0

This cutoff would exclude a significant proportion of variants within the AMR and OTH population labels. We may need to reassess the statistic that we use for this flag, but this is more of a research question and will likely take more time so we need to determine the priority of this. 

Minutes from discussion at gnomAD project meeting:
Address questions on AF discrepancies across exomes and genomes (Julia)
    - See: https://atgu.slack.com/archives/CRA2TKTV0/p1638813260070100 
    - Use case-control association framework (with population/platform covariates) to flag discrepancies in AFs
    - AMR variants are excessively flagged for exome/genome discrepancies in AF, likely due to differences in populations
    - MD: need a better-behaving statistic
    - BN: proposed model (in Slack) to try
    - MD: run just on NFE and try to tease apart contributions from exome platforms?
    - KK: expensive analysis

**Comment by gtiao** [comment]
This is now added as an agenda item for our next project meeting -- further analysis is on hold until we have a better defined test statistic to use

**Comment by gtiao** [comment]
Discussion on the appropriate test statistic to use is here: https://atgu.slack.com/archives/CNNTF8Z46/p1641242796057100

**Comment by jkgoodrich** [comment]
Note: the new test statistic requires PCs across all samples (v2 and v3). This will likely need to be run before we can perform this statistic, as I don't think this currently exists (using full set of variants, pc-relate was run with a sampling of 10% of variants).

**Comment by gtiao** [comment]
If there's a sensible set of PC loadings (probably based on the v2 exomes) we can use, let's project whichever samples don't have a consistent set of PCs onto that. Unless you think it will be quick to pick a common set of variant sites across exomes and genomes, of course.

**Comment by gtiao** [comment]
To recap discussion from the Jan 18 Project meeting, the revised test we will run is:

genome_vs_exome_status ~ genotype + [ancestry_pcs]

If we need to, we can run an additional test statistic that explicitly models which exome platform the data comes from. Mark notes the v4 exome platforms are going to look different from the v2 exome platforms, so we don't really want to over-spend resources analyzing v2 vs v3 discrepancies.

Instead, let's see if this test statistic has a better behaved distribution and could be used for deciding when to flag true technical differences between exome and genome AFs.

**Comment by jkgoodrich** [comment]
Yes loadings based on based on the v2 exomes, will work too. My comment is mostly just that currently we don't have a PCA score HT for all these samples together unless we are happy using 10% of the v2 variants. https://github.com/broadinstitute/gnomad_qc/blob/8b2342133e6186ce7056820ca5ec1b66032173bd/gnomad_qc/v3/sample_qc/v2_pc_relate.py#L29 And therefor this is an extra step in the creation of the new test statistic

**Comment by jkgoodrich** [comment]
Running the code for the logistic regression:

Cluster:
```
hailctl dataproc start jg3 --properties=spark:spark.speculation=true --autoscaling-policy=gnomad_v3_1_autoscale --max-idle=100m --requester-pays-allow-all --pkgs="git+https://github.com/broadinstitute/gnomad_methods.git@master"
```

Command: 
```
hailctl dataproc submit jg3 PycharmProjects/gnomad_qc/gnomad_qc/v3/annotations/compare_freq.py  --logistic-regression --use-pc-project --overwrite --pyfiles PycharmProjects/gnomad_qc/gnomad_qc
```

Job Link: https://console.cloud.google.com/dataproc/jobs/f79b65dd32044fd78e9bc9aab7f1c4be

**Comment by jkgoodrich** [comment]
This job failed. I ended up checkpointing the subset gnomAD v3 MT (failed with memory error, but ended up succeeding with first cluster below) and subset gnomAD v2 MT (failed with a shuffle error, succeeded with the second cluster below), and then running the logistic regression.


Cluster1: hailctl dataproc start --autoscaling-policy=gnomad_v3_1_autoscale jg2 --master-boot-disk-size 800 --worker-boot-disk-size 1000 --master-machine-type n1-highmem-16 --worker-machine-type n1-highmem-8 --max-idle=540m --requester-pays-allow-all --pkgs="git+https://github.com/broadinstitute/gnomad_methods.git@master"

Job1: 61b893158e7741d9b96663184f14a52e

Cluster2: hailctl dataproc start jg3 --master-boot-disk-size 800 --worker-boot-disk-size 1000 --master-machine-type n1-highmem-16 --worker-machine-type n1-highmem-8 --max-idle=540m --requester-pays-allow-all --pkgs="git+https://github.com/broadinstitute/gnomad_methods.git@master"

Job2: 1763e019f6104bff9c6ce3f6ad2ff7d6 

Cluster3: hailctl dataproc start --autoscaling-policy=gnomad_v3_1_autoscale jg2 --master-boot-disk-size 800 --worker-boot-disk-size 1000 --master-machine-type n1-highmem-16 --worker-machine-type n1-highmem-8 --max-idle=540m --requester-pays-allow-all --pkgs="git+https://github.com/broadinstitute/gnomad_methods.git@master"

Job3: 4c678be1eac74cd7a086d941f31c72b2


Command: hailctl dataproc submit jg1 PycharmProjects/gnomad_qc/gnomad_qc/v3/annotations/compare_freq.py  --logistic-regression --use-pc-project --overwrite --pyfiles PycharmProjects/gnomad_qc/gnomad_qc

**Comment by jkgoodrich** [comment]
Job three succeeded so I will now make the previously made plots and tables broken down by population.

**Comment by jkgoodrich** [comment]
I have updated the notebook and added the logistic regression to the spreadsheet. It look like the p-value cutoff will exclude more samples overall (we may want to reconsider the p-value cutoff). And although the AMR variants seem to be recovered more than with the chi squared p-value, other populations have a higher proportion of variants that are excluded. In all population specific plots, the variants were filtered to include only variants that are found in the given population, however the logistic regression is computed using all samples with the first 10 PCs included as covariates as described above.

https://docs.google.com/spreadsheets/d/1HtP6t2_ta4EkcyU4PIh-c9c0fUz1mhVBSlcEnC_gUqw/edit#gid=0

[https://app.zenhub.com/files/153521178/4731f752-99b8-4776-9183-daeeff3aafda/download](https://app.zenhub.com/files/153521178/4731f752-99b8-4776-9183-daeeff3aafda/download)

**Comment by jkgoodrich** [comment]
Additional discussion after ATGU/DSP meeting here: https://atgu.slack.com/archives/CRA2TKTV0/p1649184757369439

**Comment by gtiao** [comment]
tl;dr for my latest [comment](https://atgu.slack.com/archives/CRA2TKTV0/p1649884625433139?thread_ts=1649184757.369439&cid=CRA2TKTV0) on the thread: definitely wait until we have v4 native GRCh38 exomes +v3 genomes ready, and do from-scratch subpop PCAs across exomes + genomes on a per ancestry basis. The analysis will be cleaner (hopefully)

**Comment by jkgoodrich** [comment]
New ticket added for v4 run: https://app.zenhub.com/workspaces/gnomad-5f4d127ea61afc001d6be50b/issues/broadinstitute/gnomad_qc/238
